nav2_dynamic_params/src/dynamic_params_validator.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+#include <limits>
+#include "nav2_dynamic_params/dynamic_params_validator.hpp"
+
+namespace nav2_dynamic_params
+{
+
+DynamicParamsValidator::DynamicParamsValidator(rclcpp::Node::SharedPtr node, bool reject_new_params)
+: node_(node),
+  reject_new_params_(reject_new_params)
+{
+  node_->register_param_change_callback(
+    std::bind(&DynamicParamsValidator::param_validation_callback, this, std::placeholders::_1));
+}
+
+rcl_interfaces::msg::SetParametersResult DynamicParamsValidator::param_validation_callback(
+  std::vector<rclcpp::Parameter> parameters)
+{
+  auto result = rcl_interfaces::msg::SetParametersResult();
+  result.successful = true;
+
+  for (auto parameter : parameters) {
+    RCLCPP_INFO(node_->get_logger(),
+      "Parameter Change Request: %s", (parameter.get_name()).c_str());
+
+    // Check if parameter is static
+    if (check_if_static(parameter)) {
+      result.successful = false;
+      RCLCPP_WARN(node_->get_logger(),
+        "Parameter Change Denied::Parameter is static: %s", parameter.get_name().c_str());
+    }
+
+    // Validate Parameter Type
+    if (param_map_.count(parameter.get_name()) > 0) {
+      if (validate_param(parameter)) {
```

Comment:
Instead of anding together the succesfuls, just
```c++
if(!validate_param(parameter)) {
 result.successful = false;
 return result;
}
```

You could just do this everywhere and simplify the code a bit. There's no point in continuing once you've got any failure.

You could use the pattern below which is pretty common
```c++
  auto result = rcl_interfaces::msg::SetParametersResult();
  result.successful = false;

 for ((auto parameter : parameters) {
  if(!check1(parameter)
    return;
  if(!check2(parameter)
   return;
  if(!check3(parameter)
   return
  }
}
if( validation_callback)
   return validation_callback(parameters)
else
 return result;
```

Reply:
I agree we could return sooner. I suppose the only value to allowing it to run the full checks is to get more log information as to additional reasons it would fail validation.

Reply:
I've been informed that there is a precedent set in the ROS 2 code base for this method of dealing with status messages. I think the pattern I described is more common, but I'm OK with this as is.